modin/backends/pandas/query_compiler.py

```diff
@@ -1126,27 +1135,50 @@ def map_func(df):
             return result
 
         if need_reindex:
-            if isinstance(self.index, pandas.MultiIndex):
-                index_level_values = [
-                    self.index.get_level_values(lvl).unique()
-                    for lvl in np.arange(self.index.nlevels)
-                ]
-                new_index = pandas.MultiIndex.from_product(
-                    [self.columns, *index_level_values]
-                )
+            if is_correct_len:
+                if isinstance(self.index, pandas.MultiIndex):
+                    index_level_values = [
+                        self.index.get_level_values(lvl).unique()
+                        for lvl in np.arange(self.index.nlevels)
+                    ]
+                    new_index = pandas.MultiIndex.from_product(
+                        [self.columns, *index_level_values]
+                    )
+                else:
+                    if isinstance(self.columns, pandas.MultiIndex):
+                        columns_level_values = [
+                            self.columns.get_level_values(lvl).unique()
+                            for lvl in np.arange(self.columns.nlevels)
+                        ]
+                        new_index = pandas.MultiIndex.from_product(
+                            [*columns_level_values, self.index]
+                        )
+                    else:
+                        new_index = pandas.MultiIndex.from_product(
+                            [self.columns, self.index]
+                        )
             else:
                 if isinstance(self.columns, pandas.MultiIndex):
                     columns_level_values = [
                         self.columns.get_level_values(lvl).unique()
                         for lvl in np.arange(self.columns.nlevels)
                     ]
+                    index_level_values = [
+                        self.index.get_level_values(lvl).unique()
+                        for lvl in np.arange(self.index.nlevels)
+                    ]
                     new_index = pandas.MultiIndex.from_product(
-                        [*columns_level_values, self.index]
+                        [*columns_level_values, *index_level_values]
                     )
```

Comment:
there is a lot of duplicated code, can we just move that logic to a some function like this:
```python
def compute_index(index, columns, consider_index=True, consider_columns=True):
    def compute_axis(axis):
        return [
            axis.get_level_values(lvl).unique()
            for lvl in np.arange(axis.nlevels)
        ]
    new_index = compute_axis(index) if consider_index else [index]
    new_columns = compute_axis(columns) if consider_columns else [columns]
    return pandas.MultiIndex.from_product([*new_index, *new_columns])
```
and then, based on your conditions use it something like this:
```python
new_index = compute_index(self.index, self.columns, consider_index=True, consider_columns=False)
```

Reply:
fixed